File: Python/Pseudo-Palindromic Paths in a Binary Tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    """
    Time:   O(n)
    Memory: O(n)
    """

    def pseudoPalindromicPaths(self, root: TreeNode) -> int:
        return self.preorder(root, path=0)

    @classmethod
    def preorder(cls, node: Optional[TreeNode], path: int) -> int:
        if node is None:
            return 0

        path = path ^ (1 << node.val)

        if node.left is None and node.right is None:
            return int(path & (path - 1) == 0)

        return cls.preorder(node.left, path) + cls.preorder(node.right, path)

File: Java/Pseudo-Palindromic Paths in a Binary Tree.java
class Solution {

    /**
     * Time:   O(n)
     * Memory: O(n)
     */
    public int pseudoPalindromicPaths (TreeNode root) {
        return preorder(root, 0);
    }

    public int preorder(TreeNode node, int path) {
        if (node == null)
            return 0;

        path = path ^ (1 << node.val);
        if (node.left == null && node.right == null)
            if ((path & (path - 1)) == 0)
                return 1;

        return preorder(node.left, path) + preorder(node.right, path);
    }
}
